From bc8454e1d00140ad73e59f16454cc6ce90cb8e0a Mon Sep 17 00:00:00 2001
From: Frode Isaksen <frodex.isaksen@intel.com>
Date: Fri, 6 Mar 2015 17:54:43 +0100
Subject: [PATCH] bcm4356: use MSI instead of IO-APIC PCIe interrupt

Use PCIe MSI edge triggered interrupt instead of
IO-APIC level triggered interrupt.
The MSI insterrupt has 3x lower latency than interrupts
routed via the IO-APIC.

Change-Id: I9715178f01fe3a01761923304514c7b93aa0e633
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1164
Signed-off-by: Frode Isaksen <frodex.isaksen@intel.com>
---
 drivers/net/wireless/bcmdhd/dhd_pcie_linux.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/dhd_pcie_linux.c b/drivers/net/wireless/bcmdhd/dhd_pcie_linux.c
index 9da44af..f7c8568 100644
--- a/drivers/net/wireless/bcmdhd/dhd_pcie_linux.c
+++ b/drivers/net/wireless/bcmdhd/dhd_pcie_linux.c
@@ -432,13 +432,19 @@ dhdpcie_request_irq(dhdpcie_info_t *dhdpcie_info)
 {
 	dhd_bus_t *bus = dhdpcie_info->bus;
 	struct pci_dev *pdev = dhdpcie_info->bus->dev;
+	int ret;
 
+	ret = pci_enable_msi(pdev);
+	if (ret < 0) {
+		DHD_ERROR(("%s: pci_enable_msi() failed, %d, fall back to INTx\n", __FUNCTION__, ret));
+	}
 	snprintf(dhdpcie_info->pciname, sizeof(dhdpcie_info->pciname),
 	    "dhdpcie:%s", pci_name(pdev));
-	if (request_irq(pdev->irq, dhdpcie_isr, IRQF_SHARED,
-	                dhdpcie_info->pciname, bus) < 0) {
-			DHD_ERROR(("%s: request_irq() failed\n", __FUNCTION__));
-			return -1;
+	ret = request_irq(pdev->irq, dhdpcie_isr, IRQF_SHARED,
+	                dhdpcie_info->pciname, bus);
+	if (ret < 0) {
+		DHD_ERROR(("%s: request_irq() failed, %d\n", __FUNCTION__, ret));
+		return ret;
 	}
 
 	DHD_TRACE(("%s %s\n", __FUNCTION__, dhdpcie_info->pciname));
@@ -700,6 +706,7 @@ dhdpcie_free_irq(dhd_bus_t *bus)
 	if (bus) {
 		pdev = bus->dev;
 		free_irq(pdev->irq, bus);
+		pci_disable_msi(pdev);
 	}
 	DHD_TRACE(("%s: Exit\n", __FUNCTION__));
 	return;
-- 
1.9.1

